outgoingra/src/main/java/no/bankid/outgoing/ra/RaRequirements.java

```diff
@@ -6,139 +6,319 @@
-                        "wanting to add BankID app as an OTP mechanism for their BankID Netcentric users.\n\n" +
-                        "If an RA service implements these methods, than the activation of an BankID app will use " +
-                        "them to connect the BankID app with the endUser's BankID."
-        //      ,contact =
-        //      @Contact(url = "http://gigantic-server.com", name = "Fred", email = "Fred@gigagantic-server.com")
-        )
+                description = "Defines methods needed to be implemented by a Registration Authority service in banks " +
+                        "wanting to add BankID App as an OTP mechanism for their BankID Netcentric users.<p>" +
+                        "If an RA service implements these methods, than the activation of an BankID App will use " +
+                        "them to connect the BankID App with the endUser's BankID.</p>"
+        ),
+        tags = {
+                @Tag(name = "Basic RA Requirements",
+                        description = "Adds or deletes BankID App from an endUser's BankID"),
+                @Tag(name = "Activation without Code Device",
+                        description = "Activation of BankID App without no other Code Device")
+
+        },
+        servers = {
+                @Server(description = "Preprod Ra-lite",
+                        url = "https://ra-preprod.bankidnorge.no/api/enduser/bankid/netcentric/vipps/bapp")
+        }
+
 )
 
-@Tag(name = "RaRequirements",
-        description = "Administrate (BankID app OTP, Netcentric BankID) relation in endUser's bank's RA")
-@Path("/external/ra/")
-@Produces({MediaType.APPLICATION_JSON})
 
+@Path("/")
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
 public interface RaRequirements {
-    @Operation(description = "Adds BankID app to an endUser's BankID OTP mechanisms in a given bank")
+
+    String DESCRIPTION_SIGNATURE = "The Signature element, as described " +
+            "in <a href=\"https://tools.ietf.org/html/draft-cavage-http-signatures-12\">Internet-Draft - Signing HTTP Messages</a>.";
+    String EXAMPLE_SIGNATURE = "keyId=\"fa998090\",algorithm=\"rsa-sha256\"," +
+            "headers=\"(request-target) date x-client-clientname x-dataownerorgid " +
+            "x-client-requestid x-customerid\",signature=\"o7zK892....\"";
+
+    String DESCRIPTION_DATE = "The date element, the preferred format specified in " +
+            "<a href=\"https://tools.ietf.org/html/rfc7231#section-7.1.1.1\">RFC 7231 - Hypertext Transfer Protocol (HTTP/1.1): Semantics and Content</a>." +
+            "<p><b>Implementation tips:</b> Format in Java is 'EEE, dd MMM yyyy HH:mm:ss zzz', locale english.</p>";
+    String EXAMPLE_DATE = "Mon, 16 Sep 2019 12:12:21 GMT";
+
+    String DESCRIPTION_DIGEST = "hash of body, SHA-256 used" +
+            "<p><b>Implementation tips:</b>MessageDigest.getIinstance(\"SHA-256\") is not thread safe in Java</p>";
```

Comment:
```suggestion
            "<p><b>Implementation tips:</b>MessageDigest.getInstance(\"SHA-256\") is not thread safe in Java</p>";
```

I don't think this comment is correct. getInstance should definitely be thread safe, however, the returned object is stateful and not thread safe.

Reply:
Yes, getInstance() is thread-safe, it is the returned object that is not. https://stackoverflow.com/questions/17554998/need-thread-safe-messagedigest-in-java

Reply:
Of course.